sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FakeLatestOffsetSupportsTriggerAvailableNow.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.streaming.{MicroBatchStream, ReadLimit, SparkDataStream, SupportsAdmissionControl, SupportsTriggerAvailableNow}
+import org.apache.spark.sql.connector.read.streaming
+
+/**
+ * This class wraps a [[SparkDataStream]] and makes it support Trigger.AvailableNow, by
+ * overriding its [[latestOffset]] method to always return the latest offset when the method is
+ * first called. It is as if there is no new data coming in from the source after the first
+ * [[latestOffset]] call.
+ */
+class FakeLatestOffsetSupportsTriggerAvailableNow(source: SparkDataStream)
```

Comment:
can we call this `AvailableNowSourceWrapper(val delegate: SparkDataStream)`? Then you can special case this in `StreamProgress` and not need `originalSources` and things like that. You can even override the `toString` method to return the `delegate.toString` so that the metrics look correct as well

Reply:
I can do the rename here. For the places to wrap the sources let's have the discussion in the other thread.